demo/midgard-node/src/transactions/utils.ts

```diff
@@ -21,26 +21,52 @@ import { Database } from "@/services/database.js";
 export const handleSignSubmit = (
   lucid: LucidEvolution,
   signBuilder: TxSignBuilder,
-): Effect.Effect<string, Error> =>
+  onSubmitFailure: (error: SubmitError) => Effect.Effect<void>,
+  onConfirmFailure: (
+    error: ConfirmError,
+    txHash: string,
+  ) => Effect.Effect<void>,
+): Effect.Effect<string | void, Error, never> =>
   Effect.gen(function* () {
-    const signed = yield* signBuilder.sign.withWallet().completeProgram();
+    const signed = yield* signBuilder.sign
+      .withWallet()
+      .completeProgram()
+      .pipe(Effect.mapError((err) => new SignError({ err })));
     yield* Effect.logInfo("✉️  Submitting transaction...");
-    const txHash = yield* signed
-      .submitProgram()
-      .pipe(
-        Effect.retry(
-          Schedule.compose(Schedule.exponential(5_000), Schedule.recurs(5)),
-        ),
-      );
+    const txHash = yield* signed.submitProgram().pipe(
+      Effect.retry(
+        Schedule.compose(Schedule.exponential(5_000), Schedule.recurs(5)),
+      ),
+      Effect.mapError((err) => new SubmitError({ err })),
+    );
     yield* Effect.logInfo(`🚀 Transaction submitted: ${txHash}`);
     yield* Effect.logInfo(`⏳ Confirming Transaction...`);
-    yield* Effect.tryPromise(() => lucid.awaitTx(txHash, 10_000));
+    yield* Effect.tryPromise({
+      try: () => lucid.awaitTx(txHash, 10_000),
+      catch: (err: any) => new ConfirmError({ err, txHash }),
+    });
     yield* Effect.logInfo(`🎉 Transaction confirmed: ${txHash}`);
     yield* Effect.logInfo("⌛ Pausing for 10 seconds...");
     yield* Effect.sleep("10 seconds");
     yield* Effect.logInfo("✅ Pause ended.");
     return txHash;
-  });
+  }).pipe(
+    Effect.catchAll((err: HandleSignSubmitError) =>
+      Effect.gen(function* () {
+        switch (err._tag) {
+          case "SubmitError":
+            onSubmitFailure(err);
+            break;
+          case "ConfirmError":
+            onConfirmFailure(err, err.txHash);
+            break;
```

Comment:
```suggestion
            yield* onSubmitFailure(err);
            break;
          case "ConfirmError":
            yield* onConfirmFailure(err, err.txHash);
            break;
```
Since we are in an `Effect.gen` here, I think the fallbacks must also be `yield*`ed.